File: p646/p646.cpp
// Sort + Plain DP. O(n^2)
class Solution {
public:
    int findLongestChain(vector<vector<int>>& pairs) {
        int n = pairs.size();

        sort(pairs.begin(), pairs.end());

        vector<int> f(n+1, 1);

        for (int i = 1; i < n; ++i)
        {
        	for (int j = i-1; j >= 0; --j)
        		if (pairs[j][1] < pairs[i][0])
        			f[i] = max(f[i], f[j]+1);
        }

        int ans = 0;
        for (int i = 0; i < n; ++i)
        	ans = max(ans, f[i]);

        return ans;
    }
};

File: p646/p646.go
// Sort + Plain DP. O(N^2)
func findLongestChain(pairs [][]int) int {
    n := len(pairs)
    if n < 2 {
    	return n
    }

    sort.Slice(pairs, func(i, j int) bool {
    		return pairs[i][1] < pairs[j][0] ||
    			pairs[i][0] < pairs[j][0]
    	})

    f := make([]int, n)
    f[0] = 1
    for i := 1; i < n; i++ {
    	f[i] = 1
    	for j := i-1; j >= 0; j-- {
    		if pairs[i][0] > pairs[j][1] && f[i] < f[j]+1 {
    			f[i] = f[j]+1
    		}
    	}
    }

    ans := 1
    for _, v := range f {
    	if v > ans {
    		ans = v
    	}
    }
    return ans
}
